loadbalancer/algorithm.go

```diff
@@ -133,12 +132,30 @@ func withFadeIn(rnd *rand.Rand, ctx *routing.LBContext, notFadingIndexes []int,
 	case *consistentHash:
 		// If all endpoints are fading, normal consistent hash result
 		if len(notFadingIndexes) == 0 {
-			return ep[choice]
+			return choice
 		}
 		// otherwise calculate consistent hash again using endpoints which are not fading
-		return ep[a.chooseConsistentHashEndpoint(ctx, skipFadingEndpoints(notFadingIndexes))]
+		return a.chooseConsistentHashEndpoint(ctx, skipFadingEndpoints(notFadingIndexes))
+	default:
+		return choice
+	}
+}
+
+func withHealthCheck(rnd *rand.Rand, ctx *routing.LBContext, notFadingIndexes []int, choice int, algo routing.LBAlgorithm) int {
+	now := time.Now()
+	failedReqs, totalReqs := ctx.Route.LBEndpoints[choice].Metrics.GetFailedRequests()
+
+	if totalReqs == 0 || float64(failedReqs)/float64(totalReqs) < 0.95 {
```

Comment:
This behavior is ok, but looks more like a CB than a admission control behavior. I might be reading this wrong but what I understand here is: Once a backend reaches 95% error rate we just stop sending requests to it until there is no active requests to this backend, then we try again and if the backend shows apparent recover we keep sending requests to it, if it reaches 95% error rate again we stop and so on.

First I don't have big problems with this behavior but from the task description what I would expect is load being drained from this backend until we reach a point where the backend reach a equilibrium and is able to serve the requests sent to it. This is much less aggressive behavior too. And is a admission control like behavior.

Have you thought about the pros and cons of this two approaches? If yes, would you mind describing then a bit for me to understand the reasoning?

From a quick look my thinking is: in case of network partition the CB behavior would be better because it acts much faster, but in case of overload admission control seems to fare better since it will not just ditch the load of a overloaded backend to other backends that might be struggling or not. Basicaly in the overload case CB behavior could spread the failure to other backends.

If there is a case that this behaviour could be a choice for the user it might make sense to make it configurable through some flag or parameter.

Reply:
Thanks for feedback Lucas! The very first thing to do from my PoV is to split withFadeIn and withHealthCheck parts to make them not so coupled together.
But from the first glance I would like AC (not CB) to be here, because I would like some requests to go to unhealthy node just to find out the moment when it will be healthy again.

Reply:
We need to have a strategy pattern somewhere.
I think what we want is:
- probability based drops (or better re-schedule), similar to admissionControl reject calculation to allow small amount of traffic to broken endpoints to observe PHC (passive health check)
- exponential backoff